File: src/chisel/checker/factory.py
from __future__ import annotations

from dataclasses import dataclass, field

from chisel.checker.controllers.check_controller import CheckController
from chisel.checker.repositories.import_graph import ImportGraph
from chisel.checker.repositories.protocols import IImportGraph
from chisel.checker.services.app_file import AppFileService
from chisel.checker.services.complexity import ComplexityService
from chisel.checker.services.concurrency import ConcurrencyService
from chisel.checker.services.config_startup import ConfigStartupService
from chisel.checker.services.error_flow import ErrorFlowService
from chisel.checker.services.import_boundary import ImportBoundaryService
from chisel.checker.services.project_structure import ProjectStructureService
from chisel.checker.services.protocols import ICheckerService
from chisel.checker.services.session import SessionService
from chisel.checker.services.structural import StructuralService
from chisel.checker.services.suppression import SuppressionService
from chisel.checker.services.check_test_structure import CheckTestStructureService


@dataclass(slots=True)
class CheckerFactory:
    _import_graph: IImportGraph = field(default_factory=ImportGraph)
    _suppression: SuppressionService = field(default_factory=SuppressionService)

    def create_controller(self) -> CheckController:
        services: list[ICheckerService] = [
            ImportBoundaryService(_import_graph=self._import_graph),
            StructuralService(),
            ComplexityService(),
            ConcurrencyService(),
            SessionService(),
            ErrorFlowService(),
            ConfigStartupService(),
            ProjectStructureService(),
            AppFileService(),
            CheckTestStructureService(),
        ]

        return CheckController(
            _services=services,
            _suppression=self._suppression,
            _import_graph=self._import_graph,
        )

File: chisel_js/src/chisel/checker/factory.ts
import type { CheckerConfig } from "chisel/checker/config";
import type { CheckerService } from "chisel/checker/controllers/check_controller";
import type { IImportGraph } from "chisel/checker/repositories/protocols";
import { CheckController } from "chisel/checker/controllers/check_controller";
import { CheckerMode } from "chisel/checker/models/mode";
import { ImportGraph } from "chisel/checker/repositories/import_graph";
import { SuppressionService } from "chisel/checker/services/shared/suppression";
import { defaultConfig } from "chisel/checker/config";

import { ImportBoundaryService } from "chisel/checker/services/architecture/import_boundary";
import { LayoutService } from "chisel/checker/services/architecture/layout";
import { ServerLayerLeakService } from "chisel/checker/services/architecture/server_layer_leak";
import { TopologyService } from "chisel/checker/services/architecture/topology";

import { StructuralSvelteService } from "chisel/checker/services/svelte/structural";
import { ComponentEnforcementService } from "chisel/checker/services/svelte/component_enforcement";
import { ColourEnforcementService } from "chisel/checker/services/svelte/colour_enforcement";
import { ApiEndpointsService } from "chisel/checker/services/svelte/api_endpoints";
import { ErrorFlowService } from "chisel/checker/services/svelte/error_flow";
import { RouteStyleService } from "chisel/checker/services/svelte/route_style";
import { TestStructureService } from "chisel/checker/services/shared/test_structure";
import { ProjectStructureService } from "chisel/checker/services/shared/project_structure";
import { CoherenceService } from "chisel/checker/services/shared/coherence";

export interface CheckerFactoryOptions {
  readonly config?: CheckerConfig;
  readonly importGraph?: IImportGraph;
}

export class CheckerFactory {
  /**
   * Compose the checker for a project.
   *
   * The service list varies with `mode` in exactly one way: the BFF-only rules
   * (the generated API client's location, and the api-route ratio) describe a
   * topology a standalone app does not have, so they are left out rather than
   * firing nonsense. Every other rule — including the whole design-system
   * rule set — runs in both modes and cannot be turned off.
   */
  static createController(options: CheckerFactoryOptions = {}): CheckController {
    const config = options.config ?? defaultConfig();
    const importGraph = options.importGraph ?? new ImportGraph(config.tsconfig);

    const services: CheckerService[] = [
      new ImportBoundaryService(importGraph, config.mode),
      new ServerLayerLeakService(importGraph),
      new LayoutService(),
      new TopologyService(importGraph),

      new StructuralSvelteService(),
      new ComponentEnforcementService(config.designSystem.allowIn),
      new ColourEnforcementService(),
      new ErrorFlowService(),
      new TestStructureService(),
      new ProjectStructureService(),
      new CoherenceService(),
    ];

    if (config.mode === CheckerMode.BFF) {
      services.push(new ApiEndpointsService());
    } else {
      // A standalone app has remote functions; a BFF does not, so there is
      // nothing to prefer over an API route there.
      services.push(new RouteStyleService());
    }

    return new CheckController({
      services,
      config,
      importGraph,
      suppression: new SuppressionService(),
    });
  }
}
